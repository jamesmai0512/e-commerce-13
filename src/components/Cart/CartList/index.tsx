'use client'

import { Suspense, useCallback, useMemo, useState } from 'react'
import dynamic from 'next/dynamic'

// Components
const Button = dynamic(() => import('@/components/Common/Button'))
const Cart = dynamic(() => import('@/components/Cart'))

import styles from './CartList.module.css'

// type
import { CartItem } from '@/app/user/carts/page'
import Loading from '@/app/loading'
interface ICartListProps {
  carts: CartItem[]
}

const CartList = ({ carts }: ICartListProps) => {
  const [cartItems, setCartItems] = useState<CartItem[]>(carts)

  const {
    shopping_cart,
    shopping_cart_title,
    shopping_cart_info,
    shopping_cart_info_product,
    shopping_cart_list,
    shopping_cart_total,
    shopping_cart_total_content,
    shopping_cart_list_detail,
    shopping_cart_list_subtotal,
    shopping_cart_list_price,
    shopping_carts_list,
  } = styles

  const updateCartItemQuantity = useCallback(
    (cartItemId: string, newQuantity: number) => {
      if (cartItems) {
        const updatedCartItems = cartItems.map((cartItem: CartItem) =>
          cartItem.id === cartItemId
            ? { ...cartItem, quantity: newQuantity }
            : cartItem,
        )
        setCartItems(updatedCartItems)
      }
    },
    [cartItems],
  )

  const deleteCartItem = useCallback(
    (cartItemId: string) => {
      if (cartItems) {
        const updatedCartItems = cartItems.filter(
          (cartItem) => cartItem.id !== cartItemId,
        )
        setCartItems(updatedCartItems)
      }
    },
    [cartItems],
  )

  const totalPrice = useMemo(() => {
    if (cartItems) {
      return cartItems.reduce(
        (acc, cartItem) => acc + cartItem.quantity * cartItem.price,
        0,
      )
    }
  }, [cartItems])

  return (
    <Suspense fallback={<Loading />}>
      <div className={shopping_cart}>
        <h1 className={shopping_cart_title}>Your shopping cart</h1>

        <div className={shopping_cart_info}>
          <p className={shopping_cart_info_product}>Product</p>
          <p className="table_name">Quantity</p>
          <p className="table_name">Total</p>
        </div>
        <div className={shopping_cart_list}>
          {cartItems.map((cartItem) => (
            <Cart
              key={cartItem.id}
              productId={cartItem.productId}
              cartId={cartItem.id}
              quantity={cartItem.quantity}
              price={cartItem.price}
              updateCartItemQuantity={updateCartItemQuantity}
              deleteCartItem={deleteCartItem}
            />
          ))}
        </div>
        <div className={shopping_cart_total}>
          <div className={shopping_cart_total_content}>
            <div className={shopping_cart_list_detail}>
              <h3 className={shopping_cart_list_subtotal}>Subtotal</h3>
              <h2 className={shopping_cart_list_price}>Â£{totalPrice}</h2>
            </div>
            <p className={shopping_carts_list}>
              Taxes and shipping are calculated at checkout
            </p>
            <Button background="purple" text="Go to checkout" />
          </div>
        </div>
      </div>
    </Suspense>
  )
}

export default CartList
