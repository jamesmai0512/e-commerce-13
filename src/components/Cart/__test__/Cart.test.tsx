import { fireEvent, getByText, render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Cart from '../'
import mockRouter from 'next-router-mock'

jest.mock('next/router', () => require('next-router-mock'))

describe('Cart component', () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl('/carts')
  })

  const props = {
    cartId: '123',
    productId: 456,
    quantity: 2,
    price: 10.5,
    updateCartItemQuantity: jest.fn(),
    deleteCartItem: jest.fn(),
  }
  it('Should render correctly', () => {
    const component = render(<Cart {...props} />)
    expect(component).toMatchSnapshot()
  })

  it('Should render correctly', () => {
    const { getByTestId } = render(<Cart {...props} />)
    expect(getByTestId('cart-title')).toHaveTextContent('Graystone vase')
    expect(getByTestId('cart-price')).toHaveTextContent('Â£10')
  })

  it('Should increase when user click button correctly', async () => {
    const { getByRole } = render(<Cart {...props} />)

    const increaseButton = getByRole('button-control-right')
    const cartInput = getByRole('cart-input-number')

    fireEvent.click(increaseButton)
    await waitFor(() => {
      expect(cartInput).toBeInTheDocument()
    })
  })

  it('Should remove when user click button correctly', async () => {
    const { getByRole, getByText } = render(<Cart {...props} quantity={1} />)

    const decreaseButton = getByRole('button-control-left')
    const cartInput = getByRole('cart-input-number')

    fireEvent.click(decreaseButton)
    await waitFor(() => {
      expect(cartInput).toBeInTheDocument()
    })
    const confirmButon = getByText('Are you sure you want to delete?')
    fireEvent.click(confirmButon)
    await waitFor(() => {
      expect(cartInput).toBeFalsy()
    })
  })

  it('Should decrease when user click button correctly', async () => {
    const { getByRole, getByText } = render(<Cart {...props} />)

    const decreaseButton = getByRole('button-control-left')
    const cartInput = getByRole('cart-input-number')

    fireEvent.click(decreaseButton)
    await waitFor(() => {
      expect(cartInput).toBeInTheDocument()
    })
  })
})
