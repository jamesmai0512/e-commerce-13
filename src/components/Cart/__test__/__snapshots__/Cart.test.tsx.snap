// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart_layout"
        role="cart-item"
      >
        <div
          class="cart_layout_info"
        >
          <img
            alt="product-image"
            data-nimg="1"
            decoding="async"
            height="134"
            loading="eager"
            src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fjamesmai0512%2Fimage-ecommerce%2Fmain%2Fimages%2FCartItem.png&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fjamesmai0512%2Fimage-ecommerce%2Fmain%2Fimages%2FCartItem.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fjamesmai0512%2Fimage-ecommerce%2Fmain%2Fimages%2FCartItem.png&w=256&q=75 2x"
            style="color: transparent;"
            width="109"
          />
          <div
            class="cart_layout_info_product"
          >
            <h2
              class="cart_layout_info_product_title"
              data-testid="cart-title"
            >
              Graystone vase
            </h2>
            <p
              class="cart_layout_info_product_description"
            >
              A timeless ceramic vase with a tri color grey glaze.
            </p>
            <h3
              class="cart_layout_info_product_price"
              data-testid="cart-price"
            >
              £
              10.5
            </h3>
          </div>
        </div>
        <div
          class="cart_layout_info_quantity"
        >
          <div
            class="input_number_amount"
          >
            <button
              class="button_control_left"
              role="button-control-left"
            >
              -
            </button>
            <label
              for="cart-input-number"
            >
              <input
                class="input_number"
                id="cart-input-number"
                max="10"
                min="0"
                readonly=""
                role="cart-input-number"
                step="1"
                type="number"
                value="2"
              />
            </label>
            <button
              class="button_control_right"
              role="button-control-right"
            >
              +
            </button>
          </div>
        </div>
        <h3
          class="cart_layout_info_product_price_total"
        >
          £
          21
        </h3>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart_layout"
      role="cart-item"
    >
      <div
        class="cart_layout_info"
      >
        <img
          alt="product-image"
          data-nimg="1"
          decoding="async"
          height="134"
          loading="eager"
          src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fjamesmai0512%2Fimage-ecommerce%2Fmain%2Fimages%2FCartItem.png&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fjamesmai0512%2Fimage-ecommerce%2Fmain%2Fimages%2FCartItem.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fjamesmai0512%2Fimage-ecommerce%2Fmain%2Fimages%2FCartItem.png&w=256&q=75 2x"
          style="color: transparent;"
          width="109"
        />
        <div
          class="cart_layout_info_product"
        >
          <h2
            class="cart_layout_info_product_title"
            data-testid="cart-title"
          >
            Graystone vase
          </h2>
          <p
            class="cart_layout_info_product_description"
          >
            A timeless ceramic vase with a tri color grey glaze.
          </p>
          <h3
            class="cart_layout_info_product_price"
            data-testid="cart-price"
          >
            £
            10.5
          </h3>
        </div>
      </div>
      <div
        class="cart_layout_info_quantity"
      >
        <div
          class="input_number_amount"
        >
          <button
            class="button_control_left"
            role="button-control-left"
          >
            -
          </button>
          <label
            for="cart-input-number"
          >
            <input
              class="input_number"
              id="cart-input-number"
              max="10"
              min="0"
              readonly=""
              role="cart-input-number"
              step="1"
              type="number"
              value="2"
            />
          </label>
          <button
            class="button_control_right"
            role="button-control-right"
          >
            +
          </button>
        </div>
      </div>
      <h3
        class="cart_layout_info_product_price_total"
      >
        £
        21
      </h3>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
